presto-main-base/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java

```diff
@@ -725,20 +768,49 @@ private StreamProperties unordered(boolean unordered)
         {
             if (unordered) {
                 ActualProperties updatedProperties = null;
-                if (otherActualProperties != null) {
-                    updatedProperties = ActualProperties.builderFrom(otherActualProperties)
+                if (otherActualProperties.isPresent()) {
+                    updatedProperties = ActualProperties.builderFrom(otherActualProperties.get())
+                            .additionalProperty(otherActualProperties.get().getPropertiesFromUniqueColumn().map(x -> ActualProperties.builderFrom(x).unordered(true).build()))
                             .unordered(true)
                             .build();
                 }
                 return new StreamProperties(
                         distribution,
                         partitioningColumns,
                         false,
-                        updatedProperties);
+                        Optional.ofNullable(updatedProperties),
+                        streamPropertiesFromUniqueColumn.map(x -> x.unordered(true)));
+            }
+            return this;
+        }
+
+        public StreamProperties uniqueToGroupProperties()
+        {
+            if (otherActualProperties.isPresent() && otherActualProperties.get().getPropertiesFromUniqueColumn().isPresent()) {
+                if (Iterables.getOnlyElement(otherActualProperties.get().getPropertiesFromUniqueColumn().get().getLocalProperties()) instanceof UniqueProperty) {
+                    ActualProperties groupedProperties = PropertyDerivations.uniqueToGroupProperties(otherActualProperties.get().getPropertiesFromUniqueColumn().get());
+                    return new StreamProperties(distribution, partitioningColumns, ordered,
+                            otherActualProperties.map(x -> ActualProperties.builderFrom(x).additionalProperty(Optional.of(groupedProperties)).build()),
+                            streamPropertiesFromUniqueColumn.map(StreamProperties::uniqueToGroupProperties));
+                }
             }
             return this;
         }
 
+        public boolean hasUniqueProperties()
+        {
+            if (otherActualProperties.isPresent() && otherActualProperties.get().getPropertiesFromUniqueColumn().isPresent()) {
+                if (Iterables.getOnlyElement(otherActualProperties.get().getPropertiesFromUniqueColumn().get().getLocalProperties()) instanceof UniqueProperty) {
+                    ActualProperties properties = otherActualProperties.get().getPropertiesFromUniqueColumn().get();
+                    VariableReferenceExpression uniqueVariable = (VariableReferenceExpression) ((UniqueProperty) Iterables.getOnlyElement(properties.getLocalProperties())).getColumn();
+                    checkState(streamPropertiesFromUniqueColumn.isPresent() && streamPropertiesFromUniqueColumn.get().partitioningColumns.isPresent()
```

Comment:
can all of this stuff be checked in the constructor for StreamProperties instead? We want to enforce it when StreamProperties is create.  Dont' think it needs to be checked here in addition, since nothing in the method uses it.

Reply:
Sure, moved the check to the constructor